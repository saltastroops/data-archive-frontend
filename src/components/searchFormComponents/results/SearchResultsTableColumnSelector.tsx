import * as React from "react";
import styled from "styled-components";
import { LargeCheckbox } from "../../basicComponents/LargeCheckbox";
import SearchResultsTableColumn from "./ISearchResultsTableColumn";

/**
 * Properties for the search results table column selector.
 *
 * columns:
 *     The array of table columns.
 * onChange:
 *     The function to call when a checkbox is clicked. This must implement
 *     changing the column visibility. The function must expect a column's
 *     data key and the new visibility status (true for visible, false for
 *     hidden) as its arguments.
 */
interface ISearchResultsTableColumnSelectorProps {
  columns: SearchResultsTableColumn[];
  onChange: (dataKey: string, visible: boolean) => void;
}

/**
 * Responsive grid for the checkboxes.
 */
const Grid = styled.div`
  display: grid;
  grid-template-columns: 100%;
  grid-gap: 10px;
  justify-content: center;
  margin-top: 15px;

  @media screen and (min-width: 440px) {
    grid-template-columns: 190px 190px;
  }

  @media screen and (min-width: 640px) {
    grid-template-columns: 190px 190px 190px;
  }

  @media screen and (min-width: 840px) {
    grid-template-columns: 190px 190px 190px 190px;
  }

  @media screen and (min-width: 1040px) {
    grid-template-columns: 190px 190px 190px 190px 190px;
  }

  @media screen and (min-width: 1472px) {
    grid-template-columns: 190px 190px 190px 190px 190px 190px;
  }
`;

/**
 * Checkboxes for toggling the visibility of columns in the search results
 * table.
 */
class SearchResultsTableColumnSelector extends React.Component<
  ISearchResultsTableColumnSelectorProps,
  {}
> {
  render() {
    const { columns } = this.props;

    return (
      <Grid>
        {columns.map((column, index) => (
          <div key={index}>
            <label className="label">
              <LargeCheckbox
                type="checkbox"
                checked={column.visible}
                name={column.dataKey}
                onChange={this.onChange}
              />
              {column.name}
            </label>
          </div>
        ))}
      </Grid>
    );
  }

  private onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    this.props.onChange(e.target.name, e.target.checked);
  };
}

export default SearchResultsTableColumnSelector;
