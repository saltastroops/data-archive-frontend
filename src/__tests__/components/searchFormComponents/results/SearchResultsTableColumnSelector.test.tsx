import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import * as React from "react";
import SearchResultsTableColumn from "../../../../components/searchFormComponents/results/ISearchResultsTableColumn";
import SearchResultsTableColumnSelector from "../../../../components/searchFormComponents/results/SearchResultsTableColumnSelector";

const columns: SearchResultsTableColumn[] = [
  { dataKey: "proposal", name: "Proposal", visible: true },
  { dataKey: "pi", name: "Principal Investigator", visible: false },
  { dataKey: "target", name: "Target Name", visible: true }
];

describe("SearchResultsTableColumn", () => {
  it("should render correctly", () => {
    const wrapper = mount(
      <SearchResultsTableColumnSelector
        columns={columns}
        onChange={jest.fn()}
      />
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should check the correct checkboxes", () => {
    const wrapper = mount(
      <SearchResultsTableColumnSelector
        columns={columns}
        onChange={jest.fn()}
      />
    );

    expect(wrapper.find('input[name="proposal"]').prop("checked")).toBeTruthy();
    expect(wrapper.find('input[name="pi"]').prop("checked")).toBeFalsy();
    expect(wrapper.find('input[name="target"]').prop("checked")).toBeTruthy();
  });

  it("should call onChange with the correct arguments when a checkbox is clicked", () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <SearchResultsTableColumnSelector columns={columns} onChange={onChange} />
    );

    const proposalInput = wrapper.find('input[name="proposal"]');
    proposalInput.simulate("change", {
      target: { name: "proposal", checked: false }
    });

    const piInput = wrapper.find('input[name="pi"]');
    piInput.simulate("change", { target: { name: "pi", checked: true } });

    expect(onChange).toHaveBeenCalledTimes(2);
    expect(onChange).toHaveBeenNthCalledWith(1, "proposal", false);
    expect(onChange).toHaveBeenNthCalledWith(2, "pi", true);
  });
});
