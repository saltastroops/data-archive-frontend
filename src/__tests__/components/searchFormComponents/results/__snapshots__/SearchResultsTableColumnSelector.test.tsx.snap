// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResultsTableColumn should render correctly 1`] = `
.c1.c1 {
  width: 18px;
  height: 18px;
}

.c0 {
  display: grid;
  grid-template-columns: 100%;
  grid-gap: 10px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-top: 15px;
}

@media screen and (min-width:440px) {
  .c0 {
    grid-template-columns: 190px 190px;
  }
}

@media screen and (min-width:640px) {
  .c0 {
    grid-template-columns: 190px 190px 190px;
  }
}

@media screen and (min-width:840px) {
  .c0 {
    grid-template-columns: 190px 190px 190px 190px;
  }
}

@media screen and (min-width:1040px) {
  .c0 {
    grid-template-columns: 190px 190px 190px 190px 190px;
  }
}

@media screen and (min-width:1472px) {
  .c0 {
    grid-template-columns: 190px 190px 190px 190px 190px 190px;
  }
}

<SearchResultsTableColumnSelector
  columns={
    Array [
      Object {
        "dataKey": "proposal",
        "name": "Proposal",
        "visible": true,
      },
      Object {
        "dataKey": "pi",
        "name": "Principal Investigator",
        "visible": false,
      },
      Object {
        "dataKey": "target",
        "name": "Target Name",
        "visible": true,
      },
    ]
  }
  onChange={[MockFunction]}
>
  <styled.div>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bwzfXH",
            "isStatic": true,
            "lastClassName": "c0",
            "rules": Array [
              "
  display: grid;
  grid-template-columns: 100%;
  grid-gap: 10px;
  justify-content: center;
  margin-top: 15px;

  @media screen and (min-width: 440px) {
    grid-template-columns: 190px 190px;
  }

  @media screen and (min-width: 640px) {
    grid-template-columns: 190px 190px 190px;
  }

  @media screen and (min-width: 840px) {
    grid-template-columns: 190px 190px 190px 190px;
  }

  @media screen and (min-width: 1040px) {
    grid-template-columns: 190px 190px 190px 190px 190px;
  }

  @media screen and (min-width: 1472px) {
    grid-template-columns: 190px 190px 190px 190px 190px 190px;
  }
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bwzfXH",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <div
        className="c0"
      >
        <div
          key="0"
        >
          <label
            className="label"
          >
            <styled.input
              checked={true}
              name="proposal"
              onChange={[Function]}
              type="checkbox"
            >
              <StyledComponent
                checked={true}
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [
                      Object {
                        "className": "checkbox",
                        "type": "checkbox",
                      },
                    ],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-bdVaJa",
                      "isStatic": true,
                      "lastClassName": "c1",
                      "rules": Array [
                        "
  && {
    width: 18px;
    height: 18px;
  }
",
                      ],
                    },
                    "displayName": "styled.input",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-bdVaJa",
                    "target": "input",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
                name="proposal"
                onChange={[Function]}
                type="checkbox"
              >
                <input
                  checked={true}
                  className="checkbox c1"
                  name="proposal"
                  onChange={[Function]}
                  type="checkbox"
                />
              </StyledComponent>
            </styled.input>
            Proposal
          </label>
        </div>
        <div
          key="1"
        >
          <label
            className="label"
          >
            <styled.input
              checked={false}
              name="pi"
              onChange={[Function]}
              type="checkbox"
            >
              <StyledComponent
                checked={false}
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [
                      Object {
                        "className": "checkbox",
                        "type": "checkbox",
                      },
                    ],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-bdVaJa",
                      "isStatic": true,
                      "lastClassName": "c1",
                      "rules": Array [
                        "
  && {
    width: 18px;
    height: 18px;
  }
",
                      ],
                    },
                    "displayName": "styled.input",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-bdVaJa",
                    "target": "input",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
                name="pi"
                onChange={[Function]}
                type="checkbox"
              >
                <input
                  checked={false}
                  className="checkbox c1"
                  name="pi"
                  onChange={[Function]}
                  type="checkbox"
                />
              </StyledComponent>
            </styled.input>
            Principal Investigator
          </label>
        </div>
        <div
          key="2"
        >
          <label
            className="label"
          >
            <styled.input
              checked={true}
              name="target"
              onChange={[Function]}
              type="checkbox"
            >
              <StyledComponent
                checked={true}
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [
                      Object {
                        "className": "checkbox",
                        "type": "checkbox",
                      },
                    ],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-bdVaJa",
                      "isStatic": true,
                      "lastClassName": "c1",
                      "rules": Array [
                        "
  && {
    width: 18px;
    height: 18px;
  }
",
                      ],
                    },
                    "displayName": "styled.input",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-bdVaJa",
                    "target": "input",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
                name="target"
                onChange={[Function]}
                type="checkbox"
              >
                <input
                  checked={true}
                  className="checkbox c1"
                  name="target"
                  onChange={[Function]}
                  type="checkbox"
                />
              </StyledComponent>
            </styled.input>
            Target Name
          </label>
        </div>
      </div>
    </StyledComponent>
  </styled.div>
</SearchResultsTableColumnSelector>
`;
